pkg/jobs/jobrecords/job_records.go

```diff
@@ -0,0 +1,17 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package jobrecords
+
+// JobRecords is an interface for jobs.Record.
+// It is created to avoid cyclic dependency.
+type JobRecords interface {
+	ImplementsJobRecords()
+}
```

Comment:
I'm not sure what we're getting out of this. It seems like the only use of this structure is in the descs package to create a map and then inject it into the internal executor, but then, in the descs package, we never actually do anything at all with those job records. If we're never going to use the schema change job records, I don't think we've solved the core problem we were setting out to solve: how do we deal with the accumulated connExecutor state that needs to be processed before committing the transaction?

In short, if we're just going to throw away the job records, then we don't need to be creating them in the descs package and we don't need this indirection. I think what we want instead is some way to get access to the sql code which runs before committing and interacts with that state. 

I think the litmus test on all of this should be: does this change fix the bug in filetable. We need to do a bit more work to encapsulate the and mimic all of the things which happen in `sql.(*connExecutor).commitSQLTransactionInternal` inside the `descs.Txn` loop. I think that will mean further unifying the `extraTxnState` structures and providing the appropriate hooks to run the code which is run here: https://github.com/cockroachdb/cockroach/blob/4e5fac8c36d97bc4d69852fd6cc2b454c65d5c00/pkg/sql/conn_executor_exec.go#L953-L961

Reply:
> It seems like the only use of this structure is in the descs package to create a map and then inject it into the internal executor, but then, in the descs package, we never actually do anything at all with those job records. 

The primary purpose of injecting it to the IE is to make sure the job entries in this map are deleted after the txn is committed.

Currently, once the conn executor is closed, it deletes the job records when it resets its extra txn state, which I think is incorrect -- it should be controlled at the txn level, not by each child conn executor of the internal executor.

So what I have changed is that if the schema change job record is injected to the internal executor, we set the `connExecutor.extraTxnState.skipSchemaChangeRecordRelease = true`, which ensures that the entries in the map won't be deleted after closing a conn executor. 

> I think the litmus test on all of this should be: does this change fix the bug in filetable. 

I actually tried this change with [the filetable issue](https://github.com/cockroachdb/cockroach/issues/76764), see [here](https://cockroachlabs.slack.com/archives/C2C5FKPPB/p1657749955375829). It seems that with a shared schema change job records for conn executors, the invalid job record is gone, but I'm not sure if it fully solved the issue.

Reply:
Who is creating the jobs implied by the job record?

Reply:
Ah yeah you're right, I missed the `commitSQLTransactionInternal` part in`CollectionFactory.TxnWithExecutor()`. I'm thinking implement this function to make a final conn executor to commit the txn:
```go
// CommitTxn is to commit the txn bound to the internal executor.
func (ie *InternalExecutor) CommitTxn(ctx context.Context, sd *sessiondata.SessionData) error {
	if ie.extraTxnState == nil || ie.extraTxnState.txn == nil {
		return errors.New("no txn to commit")
	}
	rw := newAsyncIEResultChannel()
	stmtBuf := NewStmtBuf()
	ex, err := ie.initConnEx(ctx, rw, sd, stmtBuf, nil /* syncCallback */)
	if err != nil {
		return errors.Wrap(err, "cannot create conn executor to commit txn")
	}
	defer ex.close(ctx, externalTxnClose)

	if err := ex.commitSQLTransactionInternal(ctx); err != nil {
		return err
	}
	return nil
}
```

And in `CollectionFactory.TxnWithExecutor()` we just do
```go
	ie := cf.ieFactoryWithTxn(ctx, sd, descsCol, schemaChangeJobRecords)
	if err := f(ctx, txn, descsCol, ie); err != nil {
		return err
	}
	defer func() {
		ie.CommitTxn(ctx, sd)
	}()
```
Do you think this makes sense?

Reply:
That mostly makes sense to me except that I think the `schemaChangeJobRecords` should be created inside `ieFactoryWithTxn` and the `descs` package never needs to know anything about it. Another thing that needs to get used to commit is the `connExecutor.extraTxnState.schemaChangerState` (https://github.com/cockroachdb/cockroach/blob/dacd4ac94f7126f26ea508758bb63dd66ea35672/pkg/sql/conn_executor.go#L1336)

Please don't take this as an action item, but I'm starting wonder to what extent the work we're doing to have the `InternalExecutor` object shove things into a new connExecutor for every statement in the transaction context makes sense or whether we should just be instantiating a single `connExecutor` in the `WithTxn` case and bind them together. It's already unsafe to execute things in parallel when using the same transaction. I think if we did that, we wouldn't need this externalized `extraTxnState` struct and we'd have less opportunity to miss something important.

Reply:
> whether we should just be instantiating a single connExecutor in the WithTxn case and bind them together

Yeah, I agree. Maybe we should just have an `InternalExecutor.connExecutor` field, and use this single conn executor for all executions under the same transaction.

Reply:
> I think the schemaChangeJobRecords should be created inside ieFactoryWithTxn and the descs package never needs to know anything about it. 

Made the change in the latest commit. Also added the `schemaChangerState` field to `ie.ExtraTxnState`.